File: src/main/java/com/pretty/platform/catalog/domain/services/ProductQueryService.java
package com.pretty.platform.catalog.domain.services;

import com.pretty.platform.catalog.domain.model.queries.GetProductByIdQuery;
import com.pretty.platform.catalog.domain.model.queries.GetProductsByCategoryAndTagQuery;
import com.pretty.platform.catalog.domain.model.queries.GetProductsByNameQuery;
import com.pretty.platform.shared.domain.model.aggregates.Product;
import java.util.List;
import java.util.Optional;

/**
 * Domain service for product query operations
 */
public interface ProductQueryService {
    Optional<Product> handle(GetProductByIdQuery query);
    List<Product> handle(GetProductsByCategoryAndTagQuery query);
    List<Product> handle(GetProductsByNameQuery query);
    List<String> getAllTags();
    List<String> getAllCategories();
}

File: internal/products/domain/services/product_query_service.go
package services

import (
	"context"
	"go-service/internal/products/domain/model/entities"
	"go-service/internal/products/domain/model/queries"
)

type ProductQueryService interface {
	HandleFindByID(ctx context.Context, query queries.FindProductByIDQuery) (*entities.Product, error)
	HandleGetAll(ctx context.Context, query queries.GetAllProductsQuery) ([]*entities.Product, error)
	HandleFindByCategory(ctx context.Context, query queries.FindProductsByCategoryQuery) ([]*entities.Product, error)
}
